### ISSUE TITLE
[pcmanfm-qt] Regression : folders listed in descending order

### ISSUE THREAD

**Issue by Esclapion** [lxde/lxqt#874]
It's very annoying, limit blocking, because if I change manually for ascending order, it's not memorized and after reboot it's again descending order.

![shot-2015-11-01_00-02-05](https://cloud.githubusercontent.com/assets/3512979/10912501/175987ae-824a-11e5-9582-5fb89abe5538.jpg)

**Comment by pmattern** [comment]
A long-standing problem, indeed. Thought we had already an issue about it.

Technically maybe related to lxde/pcmanfm-qt#181.

**Comment by Esclapion** [comment]
It worked in 0.9 :-(

**Comment by excalibur1234** [comment]
this problem might be related to another (very similar) problem:
when marking multiple media files (and play them with a media player, e.g. mpv), the media files are played in descending order. despite me having them marked in ascending order.
this means, pcmanfm-qt somehow "stores" the files internally in descending order.

**Comment by tsujan** [comment]
It may have something to do with an old config file. Please
1. back up the folder '~/.config/pcmanfm-qt/',
2. close pcmanfm-qt and stop desktop from LXQT session settings,
3. remove the contents of '~/.config/pcmanfm-qt/' (with another file manager or in terminal),
4. start desktop again and see what happens.

**Comment by Esclapion** [comment]
@tsujan : absolutely no change. I only have lost my wallpaper :-)

**Comment by tsujan** [comment]
To restore the backup (including your wallpaper):
1. close pcmanfm-qt and stop desktop from LXQT session settings,
2. put your backup of the contents of  '~/.config/pcmanfm-qt/' in it (with another file manager or in terminal),
3. start desktop again.

**Comment by Esclapion** [comment]
No problem for that, I'm in VBox ;)

**Comment by tsujan** [comment]
I read the relevant parts of pcmanfm-qt's code twice and didn't find anything wrong. Should be related to another part of LXQT, IMO.

**Comment by ArthurBorsboom** [comment]
I have the same issue, since the update from 0.9 to 0.10

**Comment by tsujan** [comment]
Do you see the same issue *outside* LXDE -- I mean, when you use pcmanfm-qt under another DE? Outside LXQT, you open it, set an ascending order, then close  and reopen it and the files are arranged in a descending order?

**Comment by ArthurBorsboom** [comment]
I am sorry. I only have LXQt running on all of my desktops. I am fanatic. :)

**Comment by tsujan** [comment]
You could easily install IceWM. It's so light and good for such tests. And probably you already have an openbox session. 

You could also do as follows under LXQT:
(1) Close all instances of pcmanfm-qt, go to Config Center > Session Settings, select Desktop and click Stop;
(2) Open pcmanfm-qt and see if it remembers the ascending file order after being closed and reopened;
(3) Close it and start Desktop from Session Settings again

**Comment by ArthurBorsboom** [comment]
I have been playing around with pcmanfm-qt (killing, restarting, etc). It seems that creating a new profile has resulted in the sort order being correct and staying. One of the differences I have noticed is the file permissions on the profile.
```
arthur@pb450:~/.config/pcmanfm-qt$ ls -al lxqt_old/
total 32
drwxr-xr-x 2 arthur arthur 4096 Nov 10 23:19 .
drwxr-xr-x 5 arthur arthur 4096 Nov 10 23:25 ..
-rw------- 1 arthur arthur 1212 Nov 10 23:19 settings.conf
-rw------- 1 arthur arthur 1213 Nov  2 18:59 settings.conf.XM1415
-rw------- 1 arthur arthur 1213 Nov  3 20:14 settings.conf.XM1418
-rw------- 1 arthur arthur 1213 Nov  2 18:15 settings.conf.XM1420
-rw------- 1 arthur arthur 1212 Nov  9 00:09 settings.conf.XM1436
-rw------- 1 arthur arthur 1213 Nov  2 18:39 settings.conf.XM3499
arthur@pb450:~/.config/pcmanfm-qt$ ls -al lxqt
total 12
drwxr-xr-x 2 arthur arthur 4096 Nov 10 23:22 .
drwxr-xr-x 5 arthur arthur 4096 Nov 10 23:25 ..
-rw-r--r-- 1 arthur arthur 1210 Nov 10 23:23 settings.conf
```

I will give it a try to change the file permissions on the old profile to see if it makes a difference.

**Comment by ArthurBorsboom** [comment]
False alarm. The file permissions had nothing to do with it.

I have noticed the following.

If you change the folder sorting order (asc -> desc or the other way around), pcmanfm-qt does not write it to its settings file.

It does not write it to its settings file when rebooting, but it *does* write it when you kill the process from the command line by: kill <pid>

Maybe the reboot is so fast, that pcmanfm does not have the time to write the settings on reboot?

On the other hand, why wouldn't pcmanfm write automatically to disk within x seconds after a settings change?

**Comment by tsujan** [comment]
@ArthurBorsboom Thank you for the tests! I think I found the (hidden) cause of that and will fix it soon :)

**Comment by ArthurBorsboom** [comment]
Cool. If you have found the solution, I would like to know the result; just out of curiosity.

**Comment by tsujan** [comment]
@ the dev team
I can easily fix this issue in pcmanfm-qt but it seems rather like a *workaround* to me. Isn't lxqt-session supposed to make all running components quit before log-out/shutdown/reboot?

**Comment by palinek** [comment]
lxqt-session is stopping all (supervised) modules -> https://github.com/lxde/lxqt-session/blob/master/lxqt-session/src/lxqtmodman.cpp#L316.

**Comment by tsujan** [comment]
@palinek 
It does for me too but apparently it doesn't for some users. So, what's your opinion: should I include the workaround for such cases or should we leave it as a sign of a problem in something else (bad packaging, outdated installation, bug,...)?

**Comment by palinek** [comment]
I didn't pay much attention on this. But how is saving settings connected to a stopping an LXQt module?

If someone doesn't use `desktop` module (pcmafm-qt --desktop) then `lxqt-session` has nothing to do with pcmafm-qt at all, right?

**Comment by tsujan** [comment]
pcmanfm-qt correctly saves its settings on quitting (and not just closing its window). This issue is about reboot/logout, hence the relation to lxqt-session.
My workaround would be saving of some settings immediately but it would hide such problems.

**Comment by tsujan** [comment]
And I suppose the reporters use desktop module.

**Comment by palinek** [comment]
@tsujan I had a quick look into the pcmafm-qt code and didn't find any logic for catching the unix signals for shutdown. Can you, please, check it?
Or implement it somewhat like in liblxqt: https://github.com/lxde/liblxqt/blob/master/lxqtapplication.cpp#L116

**Comment by palinek** [comment]
Heh... just tested saving the settings.... and realized, that there are weird rights on the settings file:
```
-rw-r--r-- 1 root root 1207 2015-11-11 12:14 .config/pcmanfm-qt/default/settings.conf
```

Digged into it and figured out this:
- if I run the pcmafm-qt as root (Tool -> Open as Root) owner of the config file is changed to root:root
- that implicates no modification is possible from "normal" instances of pcmanfm-qt

IMO this is the main reason for the "lost" settings.

**Comment by palinek** [comment]
> - if I run the pcmafm-qt as root (Tool -> Open as Root) owner of the config file is changed to root:root

This is because of the internal use of the `QSaveFile` in the `QSettings` http://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/io/qsettings.cpp?h=5.5#n1472:
```c++
#ifndef QT_BOOTSTRAPPED
            QSaveFile sf(confFile->name);
#else
            QFile sf(confFile->name);
#endif
            if (!sf.open(QIODevice::WriteOnly)) {
...
```
```
The QSaveFile class provides an interface for safely writing to files.

QSaveFile is an I/O device for writing text and binary files, without losing existing data if the writing operation fails.

While writing, the contents will be written to a temporary file, and if no error happened, commit() will move it to the final file. 
```
This is definitely wrong for our use case :(

**Comment by tsujan** [comment]
> if I run the pcmafm-qt as root (Tool -> Open as Root) owner of the config file is changed to root:root

I did so twice (with and without the desktop module) and the owner didn't change. Besides, the root instance has a separate config file in the root directory.

Do you have the latest version of pcmanfm-qt? I ask because it is the only updated LXQT component I have -- the others are still from a month ago ;)

**Comment by tsujan** [comment]
In fact, when Tool -> Open as Root is used, a separate instance of pcmanfm-qt, with its own config folder in /root/.config, is launched with the su command.

**Comment by palinek** [comment]
> >if I run the pcmafm-qt as root (Tool -> Open as Root) owner of the config file is changed to root:root
>
> I did so twice (with and without the desktop module) and the owner didn't change. Besides, the root instance has a separate config file in the root directory.

What do you use for elevating the rights (Edit->Preferences->Advanced->Switch user command)?
(for me if gksu or recent version of lxqt-sudo is used.... it keeps the environment and therefore the same config file is used)

**Comment by tsujan** [comment]
I use gksu too.
I tested several times: for me, the user config files are never changed by the root instance.

**Comment by palinek** [comment]
Heh... this is weird ... Can you try it with debugs like this:
```diff
diff --git a/pcmanfm/settings.cpp b/pcmanfm/settings.cpp
index dce0d3f..9e942e9 100644
--- a/pcmanfm/settings.cpp
+++ b/pcmanfm/settings.cpp
@@ -27,6 +27,7 @@
 #include "desktopwindow.h"
 #include "utilities.h"
 // #include <QDesktopServices>
+#include <QDebug>
 
 using namespace PCManFM;
 
@@ -143,6 +144,7 @@ bool Settings::save(QString profile) {
 
 bool Settings::loadFile(QString filePath) {
   QSettings settings(filePath, QSettings::IniFormat);
+qDebug() << __FUNCTION__ << settings.fileName();
   settings.beginGroup("System");
   fallbackIconThemeName_ = settings.value("FallbackIconThemeName").toString();
   if(fallbackIconThemeName_.isEmpty()) {
@@ -242,6 +244,7 @@ bool Settings::loadFile(QString filePath) {
 
 bool Settings::saveFile(QString filePath) {
   QSettings settings(filePath, QSettings::IniFormat);
+qDebug() << __FUNCTION__ << settings.fileName();
 
   settings.beginGroup("System");
   settings.setValue("FallbackIconThemeName", fallbackIconThemeName_);
```

to see what files are read/written.

**Comment by tsujan** [comment]
In the meantime, could you please tell what's the output of echo `$XDG_CONFIG_HOME<addr>` for you **as root**? Mine is empty, so pcmanfm-qt uses the correct path `QDir::homePath() % QLatin1String("/.config");`

**Comment by tsujan** [comment]
I meant `echo $XDG_CONFIG_HOME`

**Comment by palinek** [comment]
$ gksu 'echo $XDG_CONFIG_HOME'
/home/palco/.config

**Comment by palinek** [comment]
Actualy the XDG_CONFIG_HOME is set by [startlxqt](https://github.com/lxde/lxqt-common/blob/master/startlxqt.in)... and gksu should preserve the environment. Are you starting your session in some specific way?

**Comment by tsujan** [comment]
> Are you starting your session in some specific way?

No! My display manager is LightDM and everything else is LXQT's.

**Comment by tsujan** [comment]
And those qDebug() messages show `/root/.config/pcmanfm-qt/default/settings.conf`.

**Comment by tsujan** [comment]
What if we don't use `qgetenv("XDG_CONFIG_HOME")` at all -- I mean in pcmanfm-qt? In this way we could work around any problem gksu might pose. Can you think of any side-effect?

**Comment by palinek** [comment]
I would prefer to start the root instance with -p root:
```diff
diff --git a/pcmanfm/mainwindow.cpp b/pcmanfm/mainwindow.cpp
index 38617b5..6e0e902 100644
--- a/pcmanfm/mainwindow.cpp
+++ b/pcmanfm/mainwindow.cpp
@@ -948,7 +948,9 @@ void MainWindow::on_actionOpenAsRoot_triggered() {
       QByteArray suCommand = settings.suCommand().toLocal8Bit();
       char* cmd = NULL;
       QByteArray programCommand = app->applicationFilePath().toLocal8Bit();
-      programCommand += " %U";
+      //Note: using profile root as a workaround for changing ownership of the user settings file
+      //(see https://github.com/lxde/lxqt/issues/874)
+      programCommand += " --profile root %U";
 
       if(fm_app_command_parse(suCommand.constData(), su_cmd_opts, &cmd, gpointer(programCommand.constData())) == 0) {
         /* no %s found so just append to it */
diff --git a/pcmanfm/settings.cpp b/pcmanfm/settings.cpp
index dce0d3f..394dfc8 100644
--- a/pcmanfm/settings.cpp
+++ b/pcmanfm/settings.cpp
@@ -150,7 +150,7 @@ bool Settings::loadFile(QString filePath) {
     // the value from XSETTINGS instead of hard code a fallback value.
     fallbackIconThemeName_ = "elementary"; // fallback icon theme name
   }
-  suCommand_ = settings.value("SuCommand", "gksu %s").toString();
+  suCommand_ = settings.value("SuCommand", "gksu '%s'").toString();
   setTerminal(settings.value("Terminal", "xterm").toString());
   setArchiver(settings.value("Archiver", "file-roller").toString());
   setSiUnit(settings.value("SIUnit", false).toBool());
```

**Comment by palinek** [comment]
@pmattern @ArthurBorsboom @Esclapion Can you, guys, check permissions of the setting file `$XDG_CONFIG_HOME/pcmanfm-qt/default/settings.conf`

And if they are screwed ... change it back to \<user\>:\<user\> and test. Optionaly can you test the behaviour of master and [root_profile](https://github.com/lxde/pcmanfm-qt/tree/root_profile) branch?

**Comment by tsujan** [comment]
> I would prefer to start the root instance with -p root

It would work in all cases but:
(1) Wouldn't it be weird to have a *root* file in the *user* config directory?
(2) Esclapion said that the removal of the config directory didn't have any effect. *If so,* the issue might not just be that of permissions. However, the permission IS an issue; we just don't know why gksu behaves differently for us.

**Comment by palinek** [comment]
> (1) Wouldn't it be weird to have a root file in the user config directory?

actualy, this is +1 for multiuser system... as different users can "configure their root" instances as they want and will not interfere

> However, the permission IS an issue; we just don't know why gksu behaves differently for us.

You can investigate what is the difference of environment in "normal" session and under `gksu`, `lxsu`, `lxsudo`, e.g. `gksu env`...

**Comment by Esclapion** [comment]
1) settings. conf was with owner root + 644, and it's probably bad. I changed the owner for me and rebooted, same

2)
[pat@lxqt ~]$ ls -l $XDG_CONFIG_HOME/pcmanfm-qt/default/settings.conf 
-rw-r--r-- 1 pat users 922 24 oct.  15:29 /home/pat/.config/pcmanfm-qt/default/settings.conf

[pat@lxqt ~]$ sudo echo $XDG_CONFIG_HOME
/home/pat/.config

**Comment by tsujan** [comment]
I don't have lxsu/lxsudo but the behavior of the root pcmanfm-qt is OK in whatever way I start it: it remembers its settings without affecting the user settings.

**Comment by tsujan** [comment]
> actualy, this is +1 for multiuser system... as different users can "configure their root" instances as they want and will not interfere

I don't know but, in 11 years I've used Linux, it's the first time I hear about a root config file under the user config directory!

**Comment by Esclapion** [comment]
If I run pcmanfm-qt as root (what I never do), it works. By default, it's in ascending order and it remembers the settings (i look at the date) in /root/.config/pcmanfm-qt/default/settings.conf.

If I do the same as normal user, my settings.conf is unchanged (and I stay the owner of this file).

**Comment by tsujan** [comment]
Perhaps my first guess was right: for some reason, lxqt-session may not stop pcmanfm-qt run with the desktop module. We may have 2 different problems here. @palinek : As I still use the lxqt-session from a month ago, is there any change that could cause that?

**Comment by tsujan** [comment]
> I had a quick look into the pcmafm-qt code and didn't find any logic for catching the unix signals for shutdown. Can you, please, check it?

The permission issue made me overlook that sentence. I'll surely investigate it -- maybe tomorrow.

**Comment by palinek** [comment]
A note for the gksu and preserving environment. If the backend `su` is used, environement variables are preserved, but if `sudo` is used, env variables are stripped. The default backend can be chosen by `gksu-properties`

**Comment by tsujan** [comment]
> A note for the gksu...

Excellent! Mine was sudo; I changed it to su and succeeded in giving root ownership to `~/.config/pcmanfm-qt/default/settings.conf` at last! But not only that should be seen as a bug (root config file in user config dir) but also it doesn't explain why the pcmafm-qt setting isn't saved on reboot when it's used with the desktop module because the ownership of `~/.config/pcmanfm-qt/lxqt/settings.conf` can't be changed in that way.
We're dealing with two distinct problems here.

**Comment by palinek** [comment]
You're still here :)... I think I've found the reason for not stopping all the modules by session correctly. But have no VM here to test and must finish something else. 

IMO the problem is that in lxqt-session the `logout()` method is called only upon a real logout request by a dbus call. And now I'm not sure if this happens if you choose any other (reboot, shutdown...) in `lxqt-leave` (there is fistful of providers in `liblxqt` power and LXQtProvider (which provides only the logout action) is only one of them and not the first one.... blah blah blah.. never mind.

But I'll test a patch in lxqt-session that will hopefully solve it.... stay tuned :)

**Comment by tsujan** [comment]
Good to know that :) And I'll think about fixing pcmanfm-qt's root conf file in user dir as soon as I have time.

**Comment by palinek** [comment]
@pmattern @ArthurBorsboom @Esclapion @tsujan 
The lxde/lxqt-session#59 should fix closing gracefully pcmanfm-qt => correctly saving settings on session end (reboot, shutdown) if "desktop" module is enabled.

**Comment by pmattern** [comment]
**ascending vs. descending order**
Problem is *not* related to opening PCManFM-Qt windows as root as discussed above. It can also be seen when this feature was never used before in a particular user profile.
That aside my findings here aren't exactly telling right now: if PCManFM-Qt gets launched the first time on a new user profile in LXQt the order is descending (Z → A). If PCManFM-Qt gets launched the same way in an Openbox ("only") session the order is ascending. Both times it doesn't matter whether the session is launched from a virtual terminal by `startx` or by SDDM and any adjustment made e. g. by clicking the respective column's header in "Detailed List" view (right pane) persists beyond the actual session and even rebooting.
Also, the problem cannot be seen in a window that gets launched by gksu using su as backend via "Open as Root".
All this remains the same running the commit from lxde/lxqt-session#59. So I don't know what to think about all this.

**folders and files owned by root in regular user's ~**
I can confirm Escapions findings here. If a PCManFM-Qt window gets opened via "Open as Toot" by gksu using su as backend or lxsu settings will be stored in directory/file `default/settings.conf`, both owned by root, in `/home/<user launching LXQt session>/.config/pcmanfm-qt`.
Using branch `root_profile` the said folder is named `root` instead of `default` and `gksu` doesn't work any longer, all the rest is the same. The commit proposed in lxde/pcmanfm-qt/#259 seems to do the trick, though.

On a side not gksu using sudo as backend is working flawlessly (and stores all settings in `/root`) while lxsudo just throws
![pcmanfm-qt_lxsudo-error-message](https://cloud.githubusercontent.com/assets/9860801/11152430/abf249ce-8a32-11e5-802f-cd329a6028ba.png)
both with PCManFM-Qt from latest master and lxde/pcmanfm-qt#259 / 

What I also wonder is whether there's something wrong with su because running `su -c juffed` settings get written to `~/.config/juff` of both root and regular user invoking `su`.

**Comment by tsujan** [comment]
> ... while lxsudo just throws ... both with PCManFM-Qt from latest master and lxde/pcmanfm-qt#259

I thinks that's a problem in lxsudo.

> ... running su -c juffed settings get written to ~/.config/juff of both root and regular user invoking su.

With lxde/pcmanfm-qt#259 too?

**Comment by pmattern** [comment]
> > ... while lxsudo just throws ... both with PCManFM-Qt from latest master and lxde/pcmanfm-qt#259

> I thinks that's a problem in lxsudo.

Probably, yes.

>> ... running su -c juffed settings get written to ~/.config/juff of both root and regular user invoking su.

> With lxde/pcmanfm-qt#259 too?

These findings are from a virtual terminal, totally independent of any file manager.

**Comment by tsujan** [comment]
> These findings are from a virtual terminal, totally independent of any file manager.

Oh, my bad!

**Comment by palinek** [comment]
> On a side not gksu using sudo as backend is working flawlessly (and stores all settings in /root) while lxsudo just throws

We unified the behaviour of lxqt-sudo -> preserve environment no matter what the selected backend is. So this is because of your sudoers configuration. Gksu does not preserve environment if the sudo backend is chosen.

**Comment by pmattern** [comment]
As of lxde/pcmanfm-qt@4af5a3e the settings when running instances by "Open as Root" get saved correctly using gksu both with su and sudo as backend or lxsu but not with lxsudo. This makes me wonder about the advantages of the said unification given that the behaviour in PCManFM-Qt can surely be considered a disadvantage.
The only modification in `/etc/sudoers` is line `<user> ALL=/usr/bin/pcmanfm-qt` (and a similar one for pacman). Don't know what should be wrong with it.

**Comment by Vladimir-csp** [comment]
I may not have a deep knowledge of this case, but diff this:
```
sudo sh -c 'env > /tmp/sudo-env'
gksudo "sh -c 'env > /tmp/gksudo-env'"
lxsudo sh -c 'env > /tmp/lxsudo-env'
```
IMHO, preserving such a big chunk of environment into an elevated session is **not secure**. lxsudo should preserve no more than sudo does or just reuse whatever sudo is based on.

**Comment by Esclapion** [comment]
Works for me :-1:  "Make PCManFM remember your preferred Sort Files settings
You can use View > Sort Files to change the order in which PCManFM lists the files, but PCManFM won't remember that the next time you start it. To make it remember, go to Edit > Preferences and close. That will write your current sort_type and sort_by values into ~/.config/pcmanfm/LXDE/pcmanfm.conf."

https://wiki.archlinux.org/index.php/PCManFM#Make_PCManFM_remember_your_preferred_Sort_Files_settings

**Comment by palinek** [comment]
> If I run pcmanfm-qt as root (what I never do), it works. By default, it's in ascending order and it remembers the settings (i look at the date) in /root/.config/pcmanfm-qt/default/settings.conf.

> If I do the same as normal user, my settings.conf is unchanged (and I stay the owner of this file).

The configuration file is saved if you enter/leave the "preferences" or at the end of application (that means end of session if you are using pcmafm-qt for desktop).
@Esclapion did you try "normal" way of saving after lxde/lxqt-session@302843e?

**Comment by Esclapion** [comment]
I'm not sure of the question, but i run always pcmanfm-qt as normal user, never root. The configuration is not automatically saved with my current version (Arch stable), but if I save it explicitely, it works.

**Comment by palinek** [comment]
> The configuration is not automatically saved with my current version (Arch stable),

Are you using pcmanfm-qt for desktop? If so, the settings are saved on the session end (if pcmafm-qt instance is terminated correctly). Did you try this scenario with  lxde/lxqt-session@302843e?

**Comment by Esclapion** [comment]
1) open pcmanfm-qt as normal user
2) change the order by view -> sort -> ascending
3) Menu -> Leave -> Logout
and the order is not changed.

lxqt-session : 0.10.0-2
pcmanfm-qt : 0.10.0-2

**Comment by palinek** [comment]
> lxqt-session : 0.10.0-2

I don't know what this version is based on (and either what distro are you using)... can you check if it includes the told [commit](/lxde/lxqt-session/commit/302843e) ?

**Comment by palinek** [comment]
I believe, this is fixed... please, reopen if you don't agree

**Comment by pmattern** [comment]
As of lxde/lxqt-session@c0c5c44 / lxde/pcmanfm-qt@c23f78e the sort order gets saved correctly, no matter whether it's changed by clicking on the header of column "Name" in Detailed List view or from View - Sorting.
Eventual changes get saved upon exiting the session only, but I guess this is expected behaviour.
Didn't look at any eventual issues regarding the "Open as root" stuff as this should be dealed with in an issue on its own if at all, IMO.

By default the sort order is descending which I'd consider wrong and addressed in lxde/lxqt-common/pull/48.

The sort order gets saved in section `[FolderView]` of PCManFM-Qt' `settings.ini` only. Would it eventually make sense to sync that setting with the one in section `[Desktop]`?

**Comment by tsujan** [comment]
> The sort order gets saved in section [FolderView] of PCManFM-Qt' settings.ini only. Would it eventually make sense to sync that setting with the one in section [Desktop]?

Should be fixed by https://github.com/lxde/pcmanfm-qt/pull/341.
